### PR TITLE
Online RL Optimization

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR optimizes online RL algorithm speed.

The main change is to remove the following forward. We don't need it as we can use the logprob during the actual forward. For why we can do that, see https://costa.sh/blog-understanding-why-there-isn't-a-log-probability-in-trpo-and-ppo's-objective

https://github.com/allenai/open-instruct/blob/5b5810c57df92046cf9d3b11d42611e2a7cdbabf/open_instruct/ppo_vllm_thread_ray_gtrl.py#L1137-L1139

There is also a minor optimization not to do the unnecessary `broadcast_to_vllm()` during the first iteration 

image

### CODE REVIEW

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 1]
```diff
@@ -954,14 +987,13 @@ def generate_with_engines(prompts: List[List[int]], sampling_params: SamplingPar
         )
         stats_shape = (
             args.num_epochs,
-            args.num_mini_batches * args.number_samples_per_prompt,
+            args.num_mini_batches,
```
Comment (vwxyzjn) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py:990]: This fixes 

image


to 

image

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 2]
```diff
@@ -79,6 +79,8 @@
 )
 from transformers.integrations import HfDeepSpeedConfig
 from vllm import SamplingParams
+from peft import PeftModel, get_peft_model_state_dict
```
Comment (hamishivi) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py:82]: oh re-adding peft model support? did you test this?

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 3]
```diff
@@ -634,8 +659,10 @@ def from_pretrained(
             attn_implementation="flash_attention_2",
             use_cache=False,
         )
+        self.policy_vocab_size = self.policy.config.vocab_size
         disable_dropout_in_model(self.policy)
         self.policy.gradient_checkpointing_enable()
+        # from deepspeed.ops.adam import DeepSpeedCPUAdam, FusedAdam
```
Comment (hamishivi) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py:665]: prolly remove?

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 4]
```diff
@@ -985,7 +1017,7 @@ def generate_with_engines(prompts: List[List[int]], sampling_params: SamplingPar
         queries_next = data[INPUT_IDS_PROMPT_KEY].to(device)
         ground_truths_next = data[GROUND_TRUTHS_KEY]
         datasets_next = data[DATASET_SOURCE_KEY]
-        if accelerator.is_main_process:
+        if self.rank == 0:
```
Comment (hamishivi) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py:1020]: is there a reason for this over accelerator main process?

File: open_instruct/ppo2.py
Comment (hamishivi) [inline open_instruct/ppo2.py:1]: Accidentally committed?

File: open_instruct/ppo_vllm_thread_ray_gtrl.py [hunk 5]
```diff
@@ -908,6 +943,14 @@ def repeat_generator():
             n=args.number_samples_per_prompt,
             stop=args.stop_strings,
         )
+        evaluation_generation_config = SamplingParams(
+            temperature=0.001,
```
Comment (hamishivi) [inline open_instruct/ppo_vllm_thread_ray_gtrl.py]: why do this?

File: open_instruct/grpo_vllm_thread_ray_gtrl.py (hunk 2 above)
Comment (vwxyzjn) [reply open_instruct/grpo_vllm_thread_ray_gtrl.py:82]: No did not test it, but this seems standard in openrlhf / open reaosner codebase so I just added it back.

File: open_instruct/grpo_vllm_thread_ray_gtrl.py (hunk 3 above)
Comment (vwxyzjn) [reply open_instruct/grpo_vllm_thread_ray_gtrl.py:665]: I am inclined to keep it in case we need to comment it out and test it out something (we wouldn't know where FusedAdam comes from)

File: open_instruct/grpo_vllm_thread_ray_gtrl.py (hunk 4 above)
Comment (vwxyzjn) [reply open_instruct/grpo_vllm_thread_ray_gtrl.py:1020]: The `accelerator.is_main_process` is kind of a hack. Since we now extensively using deepspeed stack, I wanted to clean it up a bit.

File: open_instruct/ppo2.py
Comment (vwxyzjn) [reply open_instruct/ppo2.py:1]: No this is intentional. ppo2.py applies the KL loss instead a KL penalty in the rewards.

File: open_instruct/ppo_vllm_thread_ray_gtrl.py (hunk 5 above)
Comment (vwxyzjn) [reply open_instruct/ppo_vllm_thread_ray_gtrl.py]: For reproducibility and better comparison? If we are using e.g., 0.7 temp, then even the same model would generate different resposnes.

Comment (hamishivi) [reply open_instruct/ppo_vllm_thread_ray_gtrl.py]: oh, but why 0.001 and not just greedy decode?

File: open_instruct/grpo_vllm_thread_ray_gtrl.py (hunk 2 above)
Comment (hamishivi) [reply open_instruct/grpo_vllm_thread_ray_gtrl.py:82]: sure, nw

**Review by hamishivi — APPROVED** [review]

File: open_instruct/ppo_vllm_thread_ray_gtrl.py (hunk 5 above)
Comment (vwxyzjn) [reply open_instruct/ppo_vllm_thread_ray_gtrl.py]: Good point. I thought we couldn't specify zero but the docs indicate otherwise. Have just changed it.

image

**Comment by billxbf** [comment]
It's a better implementation than TRL :) Great job! @vwxyzjn

**Comment by natolambert** [comment]
@vwxyzjn is a GOAT
